src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,282 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | A 'Ledger' provides access to the blockchain as well as the state
+-- and receipt associated to a block in the blockchain.
+--
+-- This module should be imported qualified.
+module Oscoin.Storage.Ledger
+    ( Ledger
+    , hoist
+    , mkLedger
+    , newMemoryIO
+    , newMemoryFromGenesisIO
+    , newFromBlockStore
+
+    , blockStoreReader
+    , blockStore
+    , getBlocksByDepth
+    , getLatestBlockWithState
+    , getTip
+    , getState
+    , getReceipt
+
+    , buildNextBlock
+    , addBlock
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain.Block
+import           Oscoin.Crypto.Blockchain.Eval
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Storage.Block.Abstract (BlockStore)
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Storage.Block.Memory
+import           Oscoin.Storage.HashStore
+import qualified Oscoin.Storage.Receipt as ReceiptStore
+import qualified Oscoin.Time.Chrono as Chrono
+
+import           Codec.Serialise (Serialise)
+import           Control.Monad.Trans.Maybe
+import qualified Crypto.Data.Auth.Tree.Class as AuthTree
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | Errors that maybe raised from the ledgers methods
+data LedgerError c
+    = StateNotInCache (BlockHash c)
+    -- ^ State for the given block is not materialized in the state
+    -- cache.
+
+deriving instance (Crypto.HasHashing c) => Eq (LedgerError c)
+deriving instance (Show (Crypto.Hash c)) => Show (LedgerError c)
+
+data Ledger crypto seal tx output state (m :: * -> *) = Ledger
+    { blockStore   :: BlockStore crypto tx seal m
+    , stateStore   :: HashStore crypto state m
+    , receiptStore :: ReceiptStore.ReceiptStore crypto tx output m
+    , ledgerEval   :: Evaluator state tx output
```

Comment:
Should this just be `eval` or `evaluator`?

Reply:
`eval` might be too ambiguous for a field name. I’ll use `ledgerEvaluator`.

Reply:
Oh no the reason I bring it up is because the other fields are not prefixed with `ledger`, but eval is. I think `ledgerEval` is fine, but thought that you might want `ledgerBlockStore` then etc.. :shrug:

Reply:
That makes sense. I prefixed all fields with `ledger` now.